Keep track of desired nodes cluster membership

fcofdez opened the pull request:
This commit adds tracking for desired nodes cluster membership.

When desired nodes are updated they are matched against the current
cluster members. Additionally when a node joins the cluster the
desired nodes cluster membership is updated.

elasticmachine commented:
Pinging @elastic/es-distributed (Team:Distributed)

elasticsearchmachine commented:
Hi @fcofdez, I've created a changelog YAML for you.

fcofdez commented on server/src/main/java/org/elasticsearch/action/admin/cluster/desirednodes/TransportUpdateDesiredNodesAction.java:
```diff
@@ -140,8 +140,13 @@ static ClusterState updateDesiredNodes(ClusterState currentState, UpdateDesiredN
             }
         }
 
+        DesiredNodesMetadata.Builder desiredNodesMetadataBuilder = new DesiredNodesMetadata.Builder(proposedDesiredNodes);
```
I think we should take into account previous desired nodes marked as members. I'll fix it.

fcofdez replied:
Tackled in https://github.com/elastic/elasticsearch/pull/84165/commits/42d504a70a8ef52a87518dae665871679048f1d9

henningandersen commented on server/src/main/java/org/elasticsearch/cluster/metadata/DesiredNodes.java:
```diff
@@ -45,21 +55,48 @@ public record DesiredNodes(String historyID, long version, List<DesiredNode> nod
+        final String historyId = in.readString();
+        final long version = in.readLong();
+        if (in.getVersion().onOrAfter(Version.CURRENT)) {
+            return new DesiredNodes(historyId, version, in.readMap(StreamInput::readString, DesiredNode::new));
```
I think we should preserve order here too? Or perhaps even switch to an ordered map (`TreeeMap`) to ensure a consistent ordering?

I wonder if we even need to change the serialization here? I think we are sending the `externalId` twice now and might as well send it as a list and then reconstruct it on the receiver side?

fcofdez replied:
I was thinking about that too. I would expect that the number of desired nodes would be fairly low in most cases. Let me revert this change.

henningandersen commented on rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/cluster.desired_nodes/10_basic.yml:
```diff
@@ -158,16 +162,49 @@ teardown:
+  - contains: { nodes: { settings: { node: { external_id: "instance-000187" } }, processors: 8, memory: "64gb", storage: "128gb", node_version: $es_version } }
+  - contains: { nodes: { settings: { node: { external_id: "instance-000188" } }, processors: 8, memory: "64gb", storage: "128gb", node_version: $es_version } }
+---
+"Test update desired nodes is idempotent with different order":
```
:+1: , but would be nice to even split this out into a separate PR? Or is the change too coupled with the rest of this PR?

fcofdez replied:
Sure, let me split this PR 👍

fcofdez replied:
I opened a new PR https://github.com/elastic/elasticsearch/pull/84227 with these changes. I think is better to hold this PR review until we get that one merged.

henningandersen reviewed (commented):
Did not get through it all, but wanted to relay a few early comments here.

fcofdez committed:
Keep track of desired nodes cluster membership

This commit adds tracking for desired nodes cluster membership.

When desired nodes are updated they are matched against the current
cluster members. Additionally when a node joins the cluster the
desired nodes cluster membership is updated.

elasticsearchmachine commented:
Hi @fcofdez, I've created a changelog YAML for you.

fcofdez committed:
Fix tests

fcofdez committed:
Update docs/changelog/84165.yaml

elasticsearchmachine commented:
Hi @fcofdez, I've created a changelog YAML for you.

henningandersen commented on server/src/main/java/org/elasticsearch/cluster/metadata/DesiredNodesMetadata.java:
```diff
@@ -21,41 +24,87 @@
+    private final Set<DesiredNode> clusterMembers;
+    private final Set<DesiredNode> notClusterMembers;
 
     public DesiredNodesMetadata(DesiredNodes latestDesiredNodes) {
```
I think we should make this private now that we have the `Builder`?

henningandersen commented on server/src/main/java/org/elasticsearch/cluster/metadata/DesiredNodesMetadata.java:
```diff
@@ -101,11 +159,76 @@ public boolean equals(Object o) {
+        private DesiredNodes desiredNodes;
+        private final Set<DesiredNode> clusterMembers;
+
+        public Builder(DesiredNodesMetadata desiredNodesMetadata) {
```
nit: We often add a `builder()` method to the outer class. I like that, since it make the client code slightly shorter and reads slightly better I think.

henningandersen reviewed (commented):
Sorry for the delay here. I wonder if we should make cluster membership a dynamically calculated property instead rather than persist it into metadata? For instance the current PR will IIUC keep all nodes as members if all nodes are abruptly shutdown and only some are started? Also at a higher level, I think knowledge about the current nodes in the cluster should not go into metadata to avoid persisting such information.

I can see a potential need for whether we ever saw a desired node, which would need to be in metadata. But I think the purpose here is different?

Left a couple random comments, not too important at this time.

fcofdez committed:
Merge remote-tracking branch 'origin/master' into desired-nodes-membership

fcofdez committed:
Revert changes

fcofdez committed:
Revert

fcofdez committed:
Merge remote-tracking branch 'origin/master' into desired-nodes-membership

fcofdez committed:
Keep track of desired nodes membership in memory

fcofdez committed:
Merge remote-tracking branch 'origin/master' into desired-nodes-membership

fcofdez committed:
Merge remote-tracking branch 'origin/master' into desired-nodes-membership

fcofdez commented:
@henningandersen I've followed a different approach where the membership is tracked in-memory, let me know if you prefer this approach. Thanks!

henningandersen commented on server/src/main/java/org/elasticsearch/cluster/metadata/DesiredNodesMembershipTracker.java:
```diff
@@ -0,0 +1,124 @@
+import java.util.HashSet;
+import java.util.Set;
+
+public class DesiredNodesMembershipTracker implements ClusterStateListener {
```
I think this needs to be a service in order for various components to be able to ask it for the members? We can also turn it into a service when we add the first use of this.

henningandersen reviewed (commented):
Thanks, this directions looks fine. I am not sure we need the quarantining/member removal though, I think we should simply remove that.

fcofdez committed:
Merge remote-tracking branch 'origin/master' into desired-nodes-membership

fcofdez committed:
Review comments

henningandersen commented on server/src/main/java/org/elasticsearch/cluster/metadata/DesiredNodesMembershipService.java:
```diff
@@ -0,0 +1,90 @@
+                    }
+                }
+
+                for (DiscoveryNode removedNode : nodesDelta.removedNodes()) {
+                    final var desiredNode = desiredNodes.find(removedNode.getExternalId());
+                    if (desiredNode != null) {
+                        members.remove(desiredNode);
+                    }
+                }
```
I think we should remove this. If we ever saw the node it is a member until a desired nodes update is made that removes it.

henningandersen commented on server/src/main/java/org/elasticsearch/cluster/metadata/DesiredNodesMembershipService.java:
```diff
@@ -0,0 +1,90 @@
+
+    @Override
+    public synchronized void clusterChanged(ClusterChangedEvent event) {
+        if (event.localNodeMaster()) {
```
I wonder if we should just maintain this always, regardless of who is master? That would ensure it is readily available on failover. Otherwise, I worry that we need another special case for the failover case, i.e., when node was not previously master, since we only consider the delta of nodes below.

fcofdez replied:
Yes, I think it should be fine if we keep track of this in all nodes and as you mention it should simplify things

henningandersen commented on server/src/test/java/org/elasticsearch/cluster/metadata/DesiredNodesMembershipServiceTests.java:
```diff
@@ -0,0 +1,291 @@
+        }
+    }
+
+    public void testMasterDemotionClearsMembers() {
```
I think we need a test demonstrating that a promoted master has the right set of members?

henningandersen reviewed (commented):
A few more things around master failover and removing nodes.

fcofdez committed:
Simplify

fcofdez committed:
Take into account more cases

fcofdez committed:
Merge remote-tracking branch 'origin/master' into desired-nodes-membership

fcofdez commented:
@elasticmachine run elasticsearch-ci/part-1 
(It was a Gradle connectivity issue)

henningandersen commented on server/src/main/java/org/elasticsearch/cluster/metadata/DesiredNodesMembershipService.java:
```diff
@@ -0,0 +1,82 @@
+        final var clusterState = event.state();
+        final var desiredNodes = DesiredNodes.latestFromClusterState(clusterState);
+        if (desiredNodes == null) {
+            final var previousDesiredNodes = DesiredNodes.latestFromClusterState(event.previousState());
+            if (previousDesiredNodes != null) {
```
Do we need this condition? Seems more expensive than just clearing `members` always?

henningandersen commented on server/src/main/java/org/elasticsearch/cluster/metadata/DesiredNodesMembershipService.java:
```diff
@@ -0,0 +1,82 @@
+    }
+
+    // visible for testing
+    synchronized int trackedMembers() {
```
```suggestion
    synchronized int trackedMembersCount() {
```

henningandersen reviewed (approved):
LGTM.

fcofdez committed:
nits

fcofdez committed:
Merge remote-tracking branch 'origin/master' into desired-nodes-membership

fcofdez commented:
@elasticmachine run elasticsearch-ci/part-2
(Unrelated LDAP failure)

fcofdez commented:
Thanks Henning!
